### PR TITLE
[Core][KV-transfer] MoRIIO: heterogeneous TP<->DP prefill/decode read routing

### PR DESCRIPTION
**Pull request by edwinlim0919**
Part of RFC #46107.

## Purpose

MoRIIO today assumes the prefill and decode engines share the same parallelism layout. This PR enables **heterogeneous** disaggregated prefill/decode, where the two phases run different parallelism (TP vs DP+EP). The performance results in the RFC which encompasses this PR shows that heterogeneous parallel PD setups can achieve higher throughput and lower TTFT at high concurrency levels than their homogeneous parallel PD counterparts on MI300X (gfx942). This PR contains the support for both mixed-parallel configurations below:

- **Forward**: `P_DP8EP → D_TP8` (DP/EP prefill, TP decode)
- **Mirror**: `P_TP8 → D_DP8EP` (TP prefill, DP/EP decode)

Three focused changes, one per commit:

### 1. Read from the request's prefill DP rank, not hardcoded DP0
Previously the READ path resolved the remote session and read-completion notify port from a hardcoded prefill DP rank 0, regardless of which prefill DP rank actually computed the KV. Under heterogeneous DP prefill the proxy forwards the owning rank in `kv_transfer_params["remote_dp_rank"]`; the read now targets it. `remote_dp_rank == 0` (the symmetric single-DP case) is byte-identical to prior behaviour.

### 2. Eager pre-forward handshake (deadlock fix)
A decode TP worker reads from whichever prefill rank owns each request, so across requests every worker must reach several remote ranks. The previous lazy per-rank handshake on the read path lets the TP workers diverge. One rank can race ahead into a forward collective (TP all-gather, or the DP+EP all-to-all on the decode side) while a peer is still blocked in a handshake `recv()` → 600s NCCL timeout / hang. We handshake every referenced remote rank **up front** in `start_load_kv`, once per engine, behind a CPU `all_reduce(MIN)` barrier so all workers finish together and fail fast + uniformly on error.

### 3. Flexible prefill-TP read (mirror)
For a TP-replicated prefill + MLA feeding a DP decode, the latent KV is replicated across every prefill TP rank, so any is a valid source. Instead of every decode DP rank funnelling onto prefill tp0 (a single-GPU NIC bottleneck), each decode rank round-robins its reads across `tp0..N-1` to spread the RDMA/prefill-NIC load. The choice is gated (all four must hold, so non-mirror configs are untouched):

- `world_size == 1`: decode is TP1, no local TP collective to keep in lockstep;
- `use_mla`: latent KV replicated across prefill TP ranks → any rank valid;
- `remote_dp_size == 1`: pure-TP prefill (no DP partition constraining source);
- `remote_tp_size > 1`: more than one prefill TP rank to spread across.

Outside the gate the source rank is the fixed local-rank mapping (`_remote_tp_rank`), byte-identical to prior behaviour. The chosen rank is threaded identically through the `(dp, tp)` session key, the handshake dial, and the notify port, so a read and its completion notify always address the **same** prefill rank. A drift would read one rank but notify another, leaving that rank's prefill buffer unfreed (KV leak → MR overflow).

The round-robin is deterministic (staggered by the decode rank's `dp_rank`), not random: it gives exactly-uniform NIC spread and is trivially testable.

**Affected files**
- `vllm/distributed/kv_transfer/kv_connector/v1/moriio/moriio_connector.py`
- `vllm/distributed/kv_transfer/kv_connector/v1/moriio/moriio_common.py`
- `tests/v1/kv_connector/unit/test_moriio_routing_fairness.py`

## Test Plan

`tests/v1/kv_connector/unit/test_moriio_routing_fairness.py`: dependency-light (no GPU/ROCm/mori), drives the **real** routing functions (`_resolve_read_source`, `_next_flex_tp_rank`) and asserts hardware-fair prefill-read distribution across
every homogeneous and mixed P/D config in the RFC, plus the flexible gate, the symmetric-TP bijection, owner-DP coverage, round-robin determinism/stagger, and the read/notify same-rank threading.

```text
Config                 coverage  balanced  flexible-gate
1P_TP8:1D_TP8          PASS      PASS      False
2P_TP8:1D_DP8EP        PASS      PASS      True
2P_TP8:2D_TP8          PASS      PASS      False
2P_DP8EP:3D_DP8EP      PASS      PASS      False
2P_DP8EP:4D_TP8        PASS      PASS      False

symmetric TP is a bijection                    PASS
owner-DP read covers every rank                PASS
round-robin deterministic/uniform/staggered    PASS
_read_blocks_for_req threads the chosen tp     PASS
21/21 assertions PASS
```

## AI assistance
This PR was prepared with AI assistance (Claude). The submitter has reviewed every
changed line and is responsible for the change end-to-end.

### CODE REVIEW

**Comment by edwinlim0919** [comment]
Tagging co-author [chaeminlim-mb](https://github.com/chaeminlim-mb) (Chaemin Lim, MangoBoost), who co-authored the fixes tracked in this issue, so he is looped in and can follow progress here.

**Comment by edwinlim0919** [comment]
Hi @Rohan138 @AndreasKaratzas @dllehr-amd @pmpalang this PR is fully tested and mergeable. The necessary context is in RFC https://github.com/vllm-project/vllm/issues/46107. Could one of you review these changes please? Thank you

**Comment by edwinlim0919** [comment]
Hi @tjtanaa @AndreasKaratzas.

This PR was validated with the testing setup below (prefill TP8, decode DP8EP), validating full accuracy with a high-concurrency gsm8k lm_eval run:

Could you please take some time to review this PR? Thank you so much.

```bash
# ---- TP8→DP8EP (strict-match 0.9613) ----
# Prefill: kv_producer, TP8
export VLLM_ROCM_USE_AITER_FUSION_SHARED_EXPERTS=0 \
python3 -m vllm.entrypoints.openai.api_server \
  --model /models/models/DeepSeek-R1 --served-model-name DeepSeek-R1-0528 --host 0.0.0.0 --port 8100 \
  --tensor-parallel-size 8 --max-model-len 20480 --max-num-batched-tokens 32768 --gpu-memory-utilization 0.80 \
  --dtype auto --trust-remote-code --distributed-executor-backend mp \
  --kv-cache-dtype fp8_e4m3 --block-size 64 --no-enable-chunked-prefill --no-enable-prefix-caching \
  --compilation-config '{"cudagraph_mode":"FULL_AND_PIECEWISE","pass_config":{"fuse_rope_kvcache":true,"fuse_rope_kvcache_cat_mla":true,"enable_sp":true,"fuse_gemm_comms":true,"fuse_act_padding":true}}' \
  --kv-transfer-config '{"kv_connector":"MoRIIOConnector","kv_role":"kv_producer","kv_connector_extra_config":{"proxy_ip":"10.1.1.182","proxy_port":36367,"proxy_ping_port":36367,"http_port":8100,"handshake_port":6301,"notify_port":61005,"read_mode":true,"qp_per_transfer":4,"pos
t_batch_size":2,"num_workers":1,"handshake_timeout":120,"max_inflight_global":16,"max_inflight_per_transfer":4,"max_dispatch_layers":4}}'

# Decode: kv_consumer, DP8EP
export VLLM_ROCM_USE_AITER_FUSION_SHARED_EXPERTS=0 AITER_BYPASS_TUNE_CONFIG=1 \
python3 -m vllm.entrypoints.openai.api_server \
  --model /models/models/DeepSeek-R1 --served-model-name DeepSeek-R1-0528 --host 0.0.0.0 --port 8201 \
  --tensor-parallel-size 1 --data-parallel-size 8 --enable-expert-parallel --all2all-backend mori_high_throughput \
  --max-model-len 20480 --max-num-batched-tokens 20480 --gpu-memory-utilization 0.80 \
  --dtype auto --trust-remote-code --distributed-executor-backend mp \
  --kv-cache-dtype fp8_e4m3 --block-size 64 --no-enable-chunked-prefill --no-enable-prefix-caching \
  --compilation-config '{"cudagraph_mode":"FULL_AND_PIECEWISE","pass_config":{"fuse_rope_kvcache":true,"fuse_rope_kvcache_cat_mla":true,"enable_sp":true,"fuse_gemm_comms":true,"fuse_act_padding":true}}' \
  --kv-transfer-config '{"kv_connector":"MoRIIOConnector","kv_role":"kv_consumer","kv_connector_extra_config":{"proxy_ip":"10.1.1.182","proxy_port":36367,"proxy_ping_port":36367,"http_port":8201,"handshake_port":6301,"notify_port":61005,"read_mode":true,"qp_per_transfer":4,"pos
t_batch_size":2,"num_workers":1,"handshake_timeout":120,"max_inflight_global":16,"max_inflight_per_transfer":4,"max_dispatch_layers":4}}'

# Proxy
python3 /app/vllm/examples/disaggregated/disaggregated_serving/moriio_toy_proxy_server.py --port 10001

# lm_eval: gsm8k, 5-shot, c512, full 1319 
python3 -m lm_eval --model local-chat-completions --apply_chat_template --tasks gsm8k --limit 1319 --log_samples \
    --model_args 'model=DeepSeek-R1-0528,base_url=http://10.1.1.182:10001/v1/chat/completions,api_key=EMPTY,num_concurrent=512,timeout=1800,tokenized_requests=False,max_length=20480' \
    --gen_kwargs max_tokens=16384,temperature=0,top_p=1
# |Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
# |-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
# |gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9621|±  |0.0053|
# |     |       |strict-match    |     5|exact_match|↑  |0.9613|±  |0.0053|
# ---- TP8→DP8EP (strict-match 0.9613) ----
```

**Comment by edwinlim0919** [comment]
Hi @tjtanaa @AndreasKaratzas the pre-commit is fixed for this PR and it has been validated with the run in the comment above. Could you please take a look at this? It is ready for review

**Comment by edwinlim0919** [comment]
hi @tjtanaa  this is one of the last PRs we require to get through. Do you giving it a review?

**Review by tjtanaa — APPROVED** [review]

**Comment by tanpinsiang** [comment]
### What was tested

| Item | Configuration |
|---|---|
| Hardware | Two AMD Instinct MI355X nodes, 8 GPUs per node |
| Model | `MiniMaxAI/MiniMax-M3-MXFP8`; identical resolved model snapshot and verified weights on both arms |
| P/D topology | TP8 prefill/router -> TP8 decode; 16 GPUs total |
| KV transfer | MoRIIO READ mode over RDMA |


### Functional and focused-test results

| Check | Baseline | PR head | Result |
|---|---:|---:|---|
| Focused MoRIIO unit suite | 59 passed | 73 passed | Pass; head includes the new routing-fairness tests |
| Router, prefill, and decode health | Healthy | Healthy | Pass |
| End-to-end P/D request | HTTP 200 | HTTP 200 | Pass |
| Router path | Both endpoints used | Both endpoints used | Pass |
| Decode transfer handshakes | All 8 TP ranks | All 8 TP ranks | Pass; head eager handshake reported `ok=True` for ranks 0-7 |
| Exception, OOM, or HTTP 5xx during valid workloads | None | None | Pass |

### Accuracy: GSM8K 25-shot

Strict exact match over all 1,319 prompts:

| Arm | Run 1 | Run 2 | Two-run mean |
|---|---:|---:|---:|
| Baseline | 94.8446% (1,251/1,319) | 94.2381% (1,243/1,319) | 94.5413% |
| PR head | 93.4799% (1,233/1,319) | 94.0864% (1,241/1,319) | 93.7832% |
| Head - baseline | -1.3647 pp | -0.1516 pp | **-0.7582 pp** |

### Performance: fixed-shape P/D benchmark


| Metric | Baseline | PR head | Head vs. baseline |
|---|---:|---:|---:|
| Successful / failed requests | 1,280 / 0 | 1,280 / 0 | Same |
| Output throughput | 3,568.583 tok/s | 3,575.756 tok/s | **+0.201%** |
| Request throughput | 3.48494 req/s | 3.49195 req/s | +0.201% |
| Total-token throughput | 32,117.245 tok/s | 32,181.803 tok/s | +0.201% |
| Duration | 367.294 s | 366.557 s | -0.201% |
| Mean TTFT | 1,850.30 ms | 1,852.04 ms | +0.094% |
| Median TTFT | 387.37 ms | 387.00 ms | -0.096% |
| p99 TTFT | 26,407.13 ms | 26,384.78 ms | -0.085% |
| Mean TPOT | 32.994 ms | 32.911 ms | **-0.253%** |
| Median TPOT | 33.440 ms | 33.382 ms | **-0.172%** |
| p99 TPOT | 33.610 ms | 33.536 ms | **-0.222%** |
| p99 ITL | 59.826 ms | 59.463 ms | -0.606% |
| Mean E2E latency | 35,603.29 ms | 35,519.65 ms | -0.235% |
| Median E2E latency | 34,613.28 ms | 34,565.19 ms | -0.139% |
| p99 E2E latency | 60,729.76 ms | 60,494.16 ms | -0.388% |

<details>
<summary>Commands used</summary>


```bash
# Router
vllm-router \
  --host 0.0.0.0 \
  --port 30000 \
  --vllm-pd-disaggregation \
  --kv-connector moriio \
  --vllm-discovery-address 0.0.0.0:36367 \
  --policy consistent_hash \
  --prefill-policy consistent_hash \
  --decode-policy consistent_hash \
  --log-level info

# Run once for the TP8 producer and once for the TP8 consumer with the
# corresponding role-specific MoRIIO configuration.
VLLM_MORIIO_CONNECTOR_READ_MODE=1 \
VLLM_USE_V1=1 \
VLLM_ROCM_USE_AITER=1 \
VLLM_USE_BREAKABLE_CUDAGRAPH=0 \
vllm serve "$MODEL_SNAPSHOT" \
  --served-model-name MiniMaxAI/MiniMax-M3-MXFP8 \
  --host 0.0.0.0 \
  --port "$ROLE_HTTP_PORT" \
  --trust-remote-code \
  --tensor-parallel-size 8 \
  --block-size 128 \
  --language-model-only \
  --kv-cache-dtype fp8 \
  --attention-backend TRITON_ATTN \
  --no-enable-prefix-caching \
  --gpu-memory-utilization 0.90 \
  --tool-call-parser minimax_m3 \
  --reasoning-parser minimax_m3 \
  --enable-auto-tool-choice \
  --kv-transfer-config "$ROLE_SPECIFIC_MORIIO_CONFIG"
```

`ROLE_SPECIFIC_MORIIO_CONFIG` selected `kv_producer` or `kv_consumer` and set `read_mode=true`, `backend=rdma`, `qp_per_transfer=4`, and `num_workers=4`.


```bash
# Run twice per arm; lm_eval==0.4.12
lm_eval run \
  --model local-completions \
  --model_args model=MiniMaxAI/MiniMax-M3-MXFP8,base_url=http://127.0.0.1:30000/v1/completions,num_concurrent=200,max_retries=10,max_gen_toks=2048,max_length=1048576,timeout=60000 \
  --tasks gsm8k \
  --num_fewshot 25 \
  --batch_size auto \
  --device cpu \
  --seed 0,1234,1234,1234 \
  --output_path "$GSM8K_OUTPUT" \
  --log_samples
```

```bash
# Warmup: NUM_PROMPTS=128 and SAVE_ARGS=().
# Measured: NUM_PROMPTS=1280 and
# SAVE_ARGS=(--save-result --save-detailed --result-dir "$PERF_OUTPUT" --result-filename result.json).
vllm bench serve \
  --backend openai \
  --host 127.0.0.1 \
  --port 30000 \
  --endpoint /v1/completions \
  --model "$MODEL_SNAPSHOT" \
  --served-model-name MiniMaxAI/MiniMax-M3-MXFP8 \
  --tokenizer "$MODEL_SNAPSHOT" \
  --dataset-name random \
  --random-input-len 8192 \
  --random-output-len 1024 \
  --random-range-ratio 0.0 \
  --num-prompts "$NUM_PROMPTS" \
  --num-warmups 0 \
  --max-concurrency 128 \
  --request-rate inf \
  --seed 0 \
  --ignore-eos \
  --trust-remote-code \
  --percentile-metrics ttft,tpot,itl,e2el \
  "${SAVE_ARGS[@]}"
```

**Comment by edwinlim0919** [comment]
@tjtanaa @tanpinsiang Thank you so much for the review, testing, and approval!!!

Currently I see that the CI fails, but the failing lanes are unrelated to this PR. The diff here is three MoRIIO files, and is unrelated to the failures.
- docker-build-cpu-image: infra failure, not a code failure (runner lost connectivity to during an apt-get build, CI node network flakiness)
- pytorch-compilation-unit-tests (exit −1): known lane timeout that is addressed by #49450
- multi-modal-processor, pipeline-plus-context-parallelism-4-gpus, v1-sample-plus-logits: inherited from main --> the six most recent commits on main (as of Jul 22) all report failing CI, and this PR has no import path into any of these subsystems

Will re-merge main and retrigger once it stabilizes (picking up #49450 in the process).
